### PR TITLE
35: onboarding device pin/biometrics not set

### LINKED ISSUE #35: If device pin/biometrics not set, provide instructions to do so (onboarding)
We need to communicate and support users to set this up without requiring device authentication to use the app.

## Requirements
* display the passcode prompt (when appropriate) on `Get Started` button tap in the welcome page
* if the user taps `Skip`, just route to FxA
* if the user hits set passcode, route to the settings page and remain on the welcome screen in the application
* don't bother storing the `Skip` preference, just launch it (when warranted) every time the user hits `Get Started`

### Engineering Notes

On iOS, this required "pausing" in driving the UI from the datastore state while onboarding. We added a `val onboarding: Observable<Boolean>` to the `RouteStore`, and ignored incoming non-onboarding routes while it was true.

Make sure that navigating to this screen prevents the user from seeing the ugly webpage error as referenced in #273

### Visual Design

**InVision**
https://mozilla.invisionapp.com/share/3SN35UHWJDX#/312603009_No_PIN

**Zeplin**
https://zpl.io/agvL3GN

Depends on #70
Depends on #30

### PR DESCRIPTION
**Pull request by eliserichards**
Fixes #35

## To Do
- [x] add “WIP” to the PR title if pushing up but not complete nor ready for review
- [x] double check the original issue to confirm it is fully satisfied
- [ ] add testing notes and screenshots in PR description to help guide reviewers
- [x] add unit tests
- [x] request the "UX" team perform a design review (if/when applicable)
- [x] make sure CI builds are passing (e.g.: fix lint and other errors)
- [x] check on the [accessibility](https://mozilla-lockbox.github.io/lockbox-android/accessibility/) of any added UI

### CODE REVIEW

File: .idea/codeStyles/Project.xml [hunk 1]
```diff
@@ -1,15 +1,5 @@
 <component name="ProjectCodeStyleConfiguration">
   <code_scheme name="Project" version="173">
-    <JetCodeStyleSettings>
```
Comment (sashei) [inline .idea/codeStyles/Project.xml]: might be nice to keep these, and the `codeStyleConfig` file that got deleted

File: .idea/misc.xml [hunk 2]
```diff
@@ -1,29 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="NullableNotNullManager">
```
Comment (sashei) [inline .idea/misc.xml]: let's keep all this project config :)

File: app/src/main/java/mozilla/lockbox/action/RouteAction.kt [hunk 3]
```diff
@@ -48,6 +48,8 @@ sealed class RouteAction(
     ) : RouteAction(TelemetryEventMethod.show, TelemetryEventObject.dialog) {
         class FingerprintDialog(@StringRes title: Int, @StringRes subtitle: Int? = null) :
             DialogFragment(dialogTitle = title, dialogSubtitle = subtitle)
+        class OnboardingSecurityDialog(@StringRes title: Int, @StringRes subtitle: Int? = null) :
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/action/RouteAction.kt]: I would use the `RouteAction.Dialog` class here, not the `DialogFragment` -- a whole fragment for this is overkill.

File: app/src/main/java/mozilla/lockbox/presenter/WelcomePresenter.kt [hunk 4]
```diff
@@ -19,15 +21,28 @@ interface WelcomeView {
 
 class WelcomePresenter(
     private val view: WelcomeView,
-    private val dispatcher: Dispatcher = Dispatcher.shared
+    private val dispatcher: Dispatcher = Dispatcher.shared,
+    private val fingerprintStore: FingerprintStore = FingerprintStore.shared
 ) : Presenter() {
     override fun onViewReady() {
+        val isDeviceSecure: Boolean = fingerprintStore.isKeyguardDeviceSecure
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/presenter/WelcomePresenter.kt]: why are you introducing a boolean to the presenter? could just be inline on ln 33.

File: .idea/misc.xml (hunk 2 above)
Comment (eliserichards) [reply .idea/misc.xml]: I think these got messed up during my re-download of AS 🤔 removing them all

File: app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt [hunk 5]
```diff
@@ -101,23 +99,24 @@ class RoutePresenter(
             is RouteAction.Dialog.SecurityDisclaimer -> showSecurityDisclaimerDialog()
             is RouteAction.Dialog.UnlinkDisclaimer -> showUnlinkDisclaimerDialog()
             is RouteAction.Dialog.NoNetworkDisclaimer -> showNoNetworkDialog()
+            is RouteAction.Dialog.OnboardingSecurityDialog -> showOnboardingSecurityDialog()
         }
 
         dialogStateObservable
-            .map { alertState ->
-                val action: Action? = when (alertState) {
+            .subscribe { alertState ->
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt]: bonus points: to keep the the minimal `dispatcher::dispatch` subscription here, you could refactor this to:
```kotlin
.map { 
    // alertState -> List<Action> 
}
.flatMapIterable { it }
.subscribe(dispatcher::dispatch)
```

Comment (sashei) [reply app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt]: reference: https://github.com/ReactiveX/RxJava/issues/2473

File: app/src/main/java/mozilla/lockbox/presenter/WelcomePresenter.kt [hunk 6]
```diff
@@ -19,15 +20,22 @@ interface WelcomeView {
...
         view.getStartedClicks
-                .subscribe { dispatcher.dispatch(RouteAction.Login) }
-                .addTo(compositeDisposable)
+            .map {
+                if (fingerprintStore.isKeyguardDeviceSecure)
+                    RouteAction.Login
+                else {
+                    RouteAction.Dialog.OnboardingSecurityDialog
+                }
+            }.subscribe(dispatcher::dispatch)
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/presenter/WelcomePresenter.kt]: nit: `\n` for the `.subscribe` call

File: app/lint-baseline.xml [hunk 7]
```diff
@@ -212,4 +212,26 @@
             column="17"/>
     </issue>
 
+    <issue
```
Comment (sashei) [inline app/lint-baseline.xml]: are these lint exclusions still required?

**Pull request merged**